### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by fahad19**
## What's done

User can now login and verify themselves:

```
$ appix login
```

^-- would open up the browser and User can sign in with their Google account.

Once signed in, User can check their login status via `appix whoami` command:

screen shot 2016-11-22 at 14 27 08

^-- shows information from DeveloperProfile service for logged in user

Docs added targetting maintainers and for developers of the ADK in README.

## Development set up

Only applicable to those developing the ADK itself.

### Environment variables

Make sure you have these environment variables (usually in your `~/.bash_profile` file) with the correct values:

```
export TRAVIX_FIREBASE_API_KEY=""
export TRAVIX_FIREBASE_AUTH_DOMAIN=""
export TRAVIX_FIREBASE_DATABASE_URL=""
export TRAVIX_FIREBASE_STORAGE_BUCKET=""
export TRAVIX_FIREBASE_MESSAGING_SENDER_ID=""

export TRAVIX_DEVELOPER_PROFILE_URL=""
```

### Build

Running this would generate `./bin/appix-mac` (if you are on macOS):

```
$ ./build.sh
```

Furthermore, you can copy it to the local appix path for easy access:

```
$ cp ./bin/appix-mac ~/.appix/appix
```

Now you can execute it as `$ appix` as usual.

## Travis-CI

Travis has been configured with the production environment variables here: https://travis-ci.org/Travix-International/Travix.Core.Adk/settings

## Possible stories in future

* Logging out
* Pass bearer token in requests to our Services requiring authorization
* Lots of clean up

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,24 +24,75 @@ $env:APPIX_VERSION='1.0.0.5'
...
+Running this would generate `./bin/appix-mac` (if you are on macOS):
+
+```
+$ ./build.sh
+```
+
+Furthermore, you can copy it to the local appix path for easy access:
+
+```
+$ cp ./bin/appix-mac ~/.appix/appix
```
Comment (alexmiranda) [inline README.md]: Or `ln -s ./bin/appix-mac ~/.appix/appix` to create a symbolic link

File: authServer.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+					.then(function (result) {
+						if (result.credential) {
+							postData('/save', {
+								content: JSON.stringify(result, null, 2)
+							});
+						}
+					})
+					.catch(function (error) {
+						console.log('error', error);
+
```
Comment (alexmiranda) [inline authServer.go]: where are not doing anything, but logging the error on the first line.

File: authServer.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+			<head>
+				` + firebaseConfig + `
+
+				<script>
+				function loginWithGoogle() {
+					firebase.auth().signInWithRedirect(provider);
+				}
+
+				function postData(path, params) {
+					var method = "post";
```
Comment (alexmiranda) [inline authServer.go]: why not `XMLHttpRequest`?

File: authServer.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+		}
+
+		formContent := r.FormValue("content")
+		content := []byte(formContent)
+		writeErr := ioutil.WriteFile(config.AuthFilePath, content, 0644)
+		if writeErr != nil {
+			panic(writeErr)
+		}
+
+		io.WriteString(w, "File written at: "+config.AuthFilePath)
```
Comment (alexmiranda) [inline authServer.go]: Should maybe inform the user they can close the browser tab?

File: whoami.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+)
+
+func configureWhoamiCommand(app *kingpin.Application) {
+	app.Command("whoami", "Displays logged in user's email address").
```
Comment (alexmiranda) [inline whoami.go]: "Displays logged in user's **information**".  What do you think?

File: authServer.go [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+	if unmarshalErr != nil {
+		log.Printf("Failed to unmarshal auth content")
+		return nil, unmarshalErr
+	}
+
+	return &auth, nil
+}
+
+func startAuthServer(c chan bool, config Config) {
```
Comment (alexmiranda) [inline authServer.go]: you can use an untyped channel and just simply send any value to it.  It's not important what gets in the channel, but that a message is sent to it.

File: login.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+	var url = "http://localhost:" + config.AuthServerPort
+
+	ch := make(chan bool)
+	go startAuthServer(ch, config)
+
+	fmt.Println("Opening url: " + url)
+	openWebsite(url)
+
+	select {
+	case shouldClose := <-ch:
```
Comment (alexmiranda) [inline login.go]: see comment about untyped channel

File: whoami.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		IdToken      string `json:"id_token"`
+		ProjectId    string `json:"project_id"`
+		RefreshToken string `json:"refresh_token"`
+		TokenType    string `json:"token_type"`
+		UserId       string `json:"user_id"`
+	}
+
+	tokenClient := &http.Client{}
+	var tokenReqPayload = []byte(`{"grant_type":"refresh_token","refresh_token": "` + refreshToken + `"}`)
+	tokenReq, _ := http.NewRequest("POST", "https://securetoken.googleapis.com/v1/token?key="+config.FirebaseApiKey, bytes.NewBuffer(tokenReqPayload))
```
Comment (alexmiranda) [inline whoami.go]: what if it errors?

File: config.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	FirebaseApiKey            string
+	FirebaseAuthDomain        string
+	FirebaseDatabaseUrl       string
+	FirebaseStorageBucket     string
+	FirebaseMessagingSenderId string
+
+	AuthServerPort string
+}
+
+func GetConfig() Config {
```
Comment (alexmiranda) [inline config.go]: yay! 👍

File: README.md (hunk 1 above)
Comment (fahad19) [inline README.md]: the `cp` example is there as a suggestion for the developer's ease. I would prefer to keep it to just copying here.

File: build.ps1 [hunk 10]
```diff
@@ -17,7 +17,7 @@ if (Test-Path env:APPVEYOR_BUILD_VERSION) {
     $GIT_HASH = "000"
 }
 $BUILD_DATE = (Get-Date -Date ((Get-Date).ToUniversalTime()) -UFormat %a.%B.%d.%Y.%R:%S) + ".+0000.UTC"
-$APP_LDFLAGS="-s -X main.version=$BUILD_VERSION -X main.gitHash=$GIT_HASH -X main.buildDate=$BUILD_DATE"
+$APP_LDFLAGS="-s -X main.version=$BUILD_VERSION -X main.gitHash=$GIT_HASH -X main.buildDate=$BUILD_DATE -X main.travixFirebaseApiKey=$TRAVIX_FIREBASE_API_KEY -X main.travixFirebaseAuthDomain=$TRAVIX_FIREBASE_AUTH_DOMAIN -X main.travixFirebaseDatabaseUrl=$TRAVIX_FIREBASE_DATABASE_URL -X main.travixFirebaseStorageBucket=$TRAVIX_FIREBASE_STORAGE_BUCKET -X main.travixFirebaseMessagingSenderId=$TRAVIX_FIREBASE_MESSAGING_SENDER_ID -X main.travixDeveloperProfileUrl=$TRAVIX_DEVELOPER_PROFILE_URL"
```
Comment (markvincze) [inline build.ps1]: Should we break this into lines?

File: build.sh [hunk 11]
```diff
@@ -5,7 +5,7 @@ set -euf -o pipefail
 BUILD_DATE=`LC_ALL=en_US.utf8 date -u +"%a.%B.%d.%Y.%R:%S.%z.%Z"`
 : "${TRAVIS_TAG:=0.0.0}"
 : "${TRAVIS_COMMIT:=`git rev-parse --short HEAD`}"
-APP_LDFLAGS="-s -X main.version=$TRAVIS_TAG -X main.gitHash=$TRAVIS_COMMIT -X main.buildDate=$BUILD_DATE"
+APP_LDFLAGS="-s -X main.version=$TRAVIS_TAG -X main.gitHash=$TRAVIS_COMMIT -X main.buildDate=$BUILD_DATE -X main.travixFirebaseApiKey=$TRAVIX_FIREBASE_API_KEY -X main.travixFirebaseAuthDomain=$TRAVIX_FIREBASE_AUTH_DOMAIN -X main.travixFirebaseDatabaseUrl=$TRAVIX_FIREBASE_DATABASE_URL -X main.travixFirebaseStorageBucket=$TRAVIX_FIREBASE_STORAGE_BUCKET -X main.travixFirebaseMessagingSenderId=$TRAVIX_FIREBASE_MESSAGING_SENDER_ID -X main.travixDeveloperProfileUrl=$TRAVIX_DEVELOPER_PROFILE_URL"
```
Comment (markvincze) [inline build.sh]: Should we break this into lines?

File: whoami.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+func executeWhoamiCommand(context *kingpin.ParseContext) error {
+	config := GetConfig()
+
+	auth, authErr := GetAuth(config)
+	if authErr != nil {
+		log.Fatal(authErr)
+		return nil
+	}
+
+	refreshToken := auth.User.StsTokenManager.RefreshToken
```
Comment (markvincze) [inline whoami.go]: Refreshing the token should probably be extracted in a way it's usable when calling the catalog in the followup TR-11442. (to me it's fine if this is only done in the followup story)

File: whoami.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+)
+
+func configureWhoamiCommand(app *kingpin.Application) {
+	app.Command("whoami", "Displays logged in user's information").
+		Action(executeWhoamiCommand)
+}
+
+func executeWhoamiCommand(context *kingpin.ParseContext) error {
```
Comment (markvincze) [inline whoami.go]: Can you break this function up in a couple of smaller functions? (maybe 2, one for refreshing the token, and another for getting the developerprofileapi response)

File: authServer.go (hunk 3 above)
Comment (mAiNiNfEcTiOn) [inline authServer.go]: yup, with `FormData` :)

**Comment by mAiNiNfEcTiOn** [comment]
So, let me see if I got this right...
We have to have those environment variables in order for `appix` to login?

That doesn't seem very logical to me...! I would've expected something like:

```
1. User downloads appix
2. User does appix login
3. User's browser opens with login form
4. User authenticates
5. We store token and refresh-token
```

Now, the only thing I'd be expecting is that - at the installation and/or first run process - the app would create the necessary environment vars.

And, _tbh_, looking at these variables,

```
export TRAVIX_FIREBASE_API_KEY=""
export TRAVIX_FIREBASE_AUTH_DOMAIN=""
export TRAVIX_FIREBASE_DATABASE_URL=""
export TRAVIX_FIREBASE_STORAGE_BUCKET=""
export TRAVIX_FIREBASE_MESSAGING_SENDER_ID=""
```

 they look like variables supposed to be in a server controlled by us, not on a client app controlled by the user...

**Comment by markvincze** [comment]
@mAiNiNfEcTiOn,
I think those env variables are used during *compilation* (see the build scripts), and not during runtime.
So the instructions about setting up those vars are for people working on developing appix and wanting to compile it locally, and not for users. For the end users, these variables are compiled into the binary. (Correct me @fahad19, if I'm wrong.)

**Comment by fahad19** [comment]
@mAiNiNfEcTiOn, @markvincze: those environment variables are only needed for the build/compilation to work. as a consumer/user of `appix`, you don't need those environment variables at all.

File: whoami.go (hunk 13 above)
Comment (fahad19) [inline whoami.go]: done.

**Comment by mAiNiNfEcTiOn** [comment]
Ok then 👍  :)

**Pull request merged**